rocksdb_admin/admin_handler.cpp

```diff
@@ -501,147 +507,317 @@ void AdminHandler::async_tm_ping(
   callback->result(BackupDBResponse());
 }
 
 void AdminHandler::async_tm_restoreDB(
     std::unique_ptr<apache::thrift::HandlerCallback<std::unique_ptr<
       RestoreDBResponse>>> callback,
     std::unique_ptr<RestoreDBRequest> request) {
-  db_admin_lock_.Lock(request->db_name);
-  SCOPE_EXIT { db_admin_lock_.Unlock(request->db_name); };
-
-  AdminException e;
+  auto full_path = FLAGS_hdfs_name_node + request->hdfs_backup_dir;
   auto upstream_addr = std::make_unique<folly::SocketAddress>();
   if (!SetAddressOrException(request->upstream_ip,
                              FLAGS_rocksdb_replicator_port,
                              upstream_addr.get(),
                              &callback)) {
+    common::Stats::get()->Incr(kHDFSRestoreFailure);
     return;
   }
 
-  auto db = db_manager_->getDB(request->db_name, nullptr);
-  if (db) {
-    e.errorCode = AdminErrorCode::DB_EXIST;
-    e.message = "Could not restore an opened DB, close it first";
-    callback.release()->exceptionInThread(std::move(e));
-    return;
-  }
-
-  auto full_path = FLAGS_hdfs_name_node + request->hdfs_backup_dir;
-  LOG(INFO) << "Restore " << request->db_name << " from " << full_path;
-
   rocksdb::Env* hdfs_env;
   auto status = rocksdb::NewHdfsEnv(&hdfs_env, full_path);
   if (!OKOrSetException(status,
                         AdminErrorCode::DB_ADMIN_ERROR,
                         &callback)) {
+    common::Stats::get()->Incr(kHDFSRestoreFailure);
     return;
   }
-  std::unique_ptr<rocksdb::Env> hdfs_env_holder(hdfs_env);
 
-  rocksdb::BackupableDBOptions options(full_path);
-  common::RocksdbGLogger logger;
-  options.info_log = &logger;
-  options.max_background_operations = FLAGS_num_hdfs_access_threads;
-  if (request->__isset.limit_mbs && request->limit_mbs > 0) {
-    options.restore_rate_limit = request->limit_mbs * kMB;
+  LOG(INFO) << "HDFS Restore " << request->db_name << " from " << full_path;
+  AdminException e;
+  if (!restoreDBHelper(request->db_name,
+                       full_path,
+                       std::unique_ptr<rocksdb::Env>(hdfs_env),
+                       std::move(upstream_addr),
+                       request->__isset.limit_mbs,
+                       request->limit_mbs,
+                       &e)) {
+    callback.release()->exceptionInThread(std::move(e));
+    common::Stats::get()->Incr(kHDFSRestoreFailure);
+    return;
   }
-  options.backup_env = hdfs_env;
 
-  rocksdb::BackupEngine* backup_engine;
-  status = rocksdb::BackupEngine::Open(
-    rocksdb::Env::Default(), options, &backup_engine);
-  if (!OKOrSetException(status,
-                        AdminErrorCode::DB_ADMIN_ERROR,
-                        &callback)) {
+  LOG(INFO) << "HDFS Restore is done.";
+  common::Stats::get()->Incr(kHDFSRestoreSuccess);
+  callback->result(RestoreDBResponse());
+}
+
+inline std::string rtrim(std::string str, char c) {
+  if (str.length() > 0 && str.back() == c) {
+    str.substr(0, str.size() - 1) ;
```

Comment:
substr() won't change the content of str.

can we do the following:

`if (!str.empty() && str.back() == c) {
    str.pop_back();
}`

Reply:
good catch!